### PR TITLE
WIP - selenium tests skeleton

### PR DESCRIPTION
**Pull request by rhopp**
### What does this PR do?
Adds skeleton of functional tests

### What issues does this PR fix or reference?
https://github.com/redhat-developer/che-functional-tests/issues/286

Followup of Tibors PR: https://github.com/redhat-developer/rh-che/pull/721

### CODE REVIEW

**Comment by rhopp** [comment]
Ilya's comment in previous PR:
> Looks good as an initial contribution +1 (the tests could be cleaner IMO, but in general it is a great improvement)

Yes... This PR could be much cleaner. But I would prefer to merge it in this state. We have plans for simplifying that by, for example, not starting che-starter in code, but use docker image as suggested by Alex. And we will address these improvement in subsequent PRs to have the smaller change sets instead one huge PR.

File: functional-tests/pom.xml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		</dependency>
+		<dependency>
+			<groupId>org.eclipse.jgit</groupId>
+			<artifactId>org.eclipse.jgit</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.jboss.shrinkwrap.resolver</groupId>
+			<artifactId>shrinkwrap-resolver-api-maven-embedded</artifactId>
+		</dependency>
+		<dependency>
```
Comment (davidfestal) [inline functional-tests/pom.xml]: According to the build, it seems you should remove this dependency that is unused and not necessary: 

```
INFO] --- maven-dependency-plugin:3.0.1:analyze-only (analyze) @ functional-tests ---
[WARNING] Unused declared dependencies found:
[WARNING]    org.jboss.shrinkwrap.resolver:shrinkwrap-resolver-impl-maven-embedded:jar:3.1.3:compile
```

**Comment by rhopp** [comment]
@davidfestal And that's the problem. When I remove this (supposedly) unused dependency, the maven build passes, but the tests fail with this message:
```
1) Error in custom provider, org.jboss.shrinkwrap.resolver.api.Invokable$InvocationException: Unable to invoke onlyOne([Ljava.lang.Class;@24a298a6) on object org.jboss.shrinkwrap.resolver.spi.loader.ServiceRegistry with parameters [Ljava.lang.Object;@982bb90
  at org.eclipse.che.selenium.core.CheSeleniumSuiteModule.getWorkspace(CheSeleniumSuiteModule.java:156) (via modules: com.google.inject.util.Modules$OverrideModule -> org.eclipse.che.selenium.core.CheSeleniumSuiteModule)
  while locating org.eclipse.che.selenium.core.workspace.TestWorkspace
    for field at com.redhat.che.functional.tests.TestTestClass.testWorkspace(TestTestClass.java:23)
  while locating com.redhat.che.functional.tests.TestTestClass

1 error
	at com.google.inject.internal.Errors.throwProvisionExceptionIfErrorsExist(Errors.java:486)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:67)
	at com.google.inject.internal.InjectorImpl.injectMembers(InjectorImpl.java:987)
	at org.eclipse.che.selenium.core.inject.SeleniumTestHandler.injectDependencies(SeleniumTestHandler.java:278)
	at org.eclipse.che.selenium.core.inject.SeleniumTestHandler.beforeInvocation(SeleniumTestHandler.java:242)
	... 22 more
Caused by: org.jboss.shrinkwrap.resolver.api.Invokable$InvocationException: Unable to invoke onlyOne([Ljava.lang.Class;@24a298a6) on object org.jboss.shrinkwrap.resolver.spi.loader.ServiceRegistry with parameters [Ljava.lang.Object;@982bb90
	at org.jboss.shrinkwrap.resolver.api.Invokable.invokeMethod(Invokable.java:109)
	at org.jboss.shrinkwrap.resolver.api.ResolverSystemFactory.createFromUserView(ResolverSystemFactory.java:93)
	at org.jboss.shrinkwrap.resolver.api.ResolverSystemFactory.createFromUserView(ResolverSystemFactory.java:54)
	at org.jboss.shrinkwrap.resolver.api.Resolvers.use(Resolvers.java:75)
	at org.jboss.shrinkwrap.resolver.api.maven.embedded.EmbeddedMaven.forProject(EmbeddedMaven.java:50)
	at org.jboss.shrinkwrap.resolver.api.maven.embedded.EmbeddedMaven.forProject(EmbeddedMaven.java:69)
	at com.redhat.che.selenium.core.workspace.CheStarterWrapper.start(CheStarterWrapper.java:88)
	at com.redhat.che.selenium.core.workspace.RhCheTestWorkspaceProvider.createWorkspace(RhCheTestWorkspaceProvider.java:59)
	at org.eclipse.che.selenium.core.CheSeleniumSuiteModule.getWorkspace(CheSeleniumSuiteModule.java:156)
	at org.eclipse.che.selenium.core.CheSeleniumSuiteModule$$FastClassByGuice$$87827f64.invoke(<generated>)
	at com.google.inject.internal.ProviderMethod$FastClassProviderMethod.doProvision(ProviderMethod.java:264)
	at com.google.inject.internal.ProviderMethod$Factory.provision(ProviderMethod.java:401)
	at com.google.inject.internal.ProviderMethod$Factory.get(ProviderMethod.java:376)
	at com.google.inject.internal.SingleFieldInjector.inject(SingleFieldInjector.java:54)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:132)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:93)
	at com.google.inject.internal.MembersInjectorImpl$1.call(MembersInjectorImpl.java:80)
	at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1085)
	at com.google.inject.internal.MembersInjectorImpl.injectAndNotify(MembersInjectorImpl.java:80)
	at com.google.inject.internal.MembersInjectorImpl.injectMembers(MembersInjectorImpl.java:62)
	... 25 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.shrinkwrap.resolver.api.Invokable.invokeMethod(Invokable.java:100)
	... 44 more
Caused by: java.lang.IllegalStateException: There are no services for serviceClass org.jboss.shrinkwrap.resolver.api.maven.embedded.pom.equipped.PomUnequippedEmbeddedMaven
	at org.jboss.shrinkwrap.resolver.spi.loader.SpiServiceLoader.onlyOne(SpiServiceLoader.java:88)
	at org.jboss.shrinkwrap.resolver.spi.loader.ServiceRegistry.onlyOne(ServiceRegistry.java:102)
	... 49 more
```
I'll try to ask shrinkwrap commiter if he encountered something like this.

**Comment by rhopp** [comment]
The PR check tests failed, probably because of https://bugzilla.redhat.com/show_bug.cgi?id=1524305 (you can see in the logs the failure of annotating route)
```
The Route "rhche" is invalid: metadata.annotations: Invalid value: "kubernetes.io~1tls-acme": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
```

The build is finally successful though. @ibuziuk If you would be so kind and merge this, please ;-)

**Comment by rhopp** [comment]
Rebased on master, build is OK. @ibuziuk pls merge.

**Pull request merged**